### PR TITLE
[stdlib] Use ephemeral string for substring comparison

### PR DESCRIPTION
**Pull request by airspeedswift**
To avoid unnecessary copies. Also makes StringProtocol comparable and implements comparisons generically.

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Review by moiseev — APPROVED** [review]
LGTM

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

**Comment by airspeedswift** [comment]
Oops, defining heterogenous `==` doesn't get us `!=` for free. Nor the other operators like `<=`, `>` etc.

**Comment by airspeedswift** [comment]
Have fixed `!=` but I guess we'll have to define all of the other comparison operators as well.

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by moiseev** [comment]
@airspeedswift if we had a `ComparableTo<T>`, then the derived operators could be possible in an extension.

**Comment by airspeedswift** [comment]
Gah and skin tone emoji aren't working right on our Linux version of ICU so I need to adjust the test.

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test

File: stdlib/public/core/Substring.swift.gyb [hunk 1]
```diff
@@ -317,42 +317,25 @@ extension Substring : LosslessStringConvertible {
   }
 }
 
-extension Substring : Equatable {
-  public static func ==(lhs: Substring, rhs: Substring) -> Bool {
-    return String(lhs) == String(rhs)
+extension StringProtocol {
+  public static func ==<S: StringProtocol>(lhs: Self, rhs: S) -> Bool {
```
Comment (dabrahams) [inline stdlib/public/core/Substring.swift.gyb]: `R` or `RHS` would probably be a better generic parameter name.  There are enough `S`'s flying around here already

**Review by dabrahams — CHANGES_REQUESTED** [review]
If there are interesting speedups to be observed here it's most likely only through the `Equatable` and `Comparable` conformances, e.g.  `arrayOfStrings.index(of: someString)`.  You might want to make sure those cases are checked in the benchmarks

**Comment by airspeedswift** [comment]
@swift-ci please smoke test

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
@swift-ci please smoke test

**Comment by airspeedswift** [comment]
@swift-ci please smoke test

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test

**Pull request merged**